flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -1277,6 +1284,37 @@ private void discardCheckpoint(PendingCheckpoint pendingCheckpoint, @Nullable Th
 		}
 	}
 
+	private void tryHandleCheckpointException(DeclineCheckpoint message, PendingCheckpoint currentCheckpoint) {
+		boolean needFailJob = message.getReason() != null && !(message.getReason() instanceof CheckpointDeclineException)
+			&& failOnCheckpointingErrors && !currentCheckpoint.getProps().isSavepoint();
+
+		if (needFailJob) {
+			boolean processed = false;
+			ExecutionAttemptID failedExecutionID = message.getTaskExecutionId();
+			for (ExecutionVertex vertex : tasksToWaitFor) {
+				if (vertex.getCurrentExecutionAttempt().getAttemptId().equals(failedExecutionID)) {
+					if (currentPeriodicTrigger != null) {
+						currentPeriodicTrigger.cancel(true);
+						currentPeriodicTrigger = null;
+					}
+
+					vertex.fail(message.getReason());
+					processed = true;
+					break;
+				}
+			}
+
+			if (!processed) {
```

Comment:
In which case will `processed` be false, and should we throw an Exception in that case or just log error is ok?

Reply:
When we can't find the ExecutionVertex corresponding to the failed task, it is false. This judgment is more for defensive purposes.

Reply:
In which case we could not find the ExecutionVertex corresponding to the failed task? In `CheckpointCoordinator::receiveDeclineMessage` we have checked the jobid, I think this `processed` will always be true, am I right?

Reply:
I believe that you are right in most cases, but there is a possibility that the network has a very serious delay. At this time, the job has failed over, and the attempt id has changed. Although this is unlikely, it is a defensive programming mechanism.